File: algorithms-python/list/nth_node.py
from typing import Optional

from list.linked_list import Node, LinkedList


def find_node_from_end(node: Node, n: int) -> Optional[Node]:
    if node is None:
        return None
    i: int = 0
    temp: Node = node
    while i < n - 1:
        temp = temp.next
        i += 1
    nth_node: Optional[Node] = None
    while temp is not None:
        if nth_node is not None:
            nth_node = nth_node.next
        else:
            nth_node = node
        temp = temp.next
    return nth_node


def display(node: Node):
    if node is None:
        return
    temp: Node = node
    while temp is not None:
        print(temp.data, end="\t")
        temp = temp.next


if __name__ == '__main__':
    linked_list: LinkedList = LinkedList()
    linked_list.add_last(Node(1))
    linked_list.add_last(Node(2))
    linked_list.add_last(Node(3))
    linked_list.add_last(Node(4))
    linked_list.add_last(Node(5))
    linked_list.add_last(Node(6))
    linked_list.add_last(Node(7))
    head: Node = linked_list.head
    display(head)
    location: int = 2
    print(f"\n{location} th node from end = {find_node_from_end(head, location).data}")

File: algortihms-java/src/com/algo/linkedList/NthNode.java
package com.algo.linkedList;

//Problem: https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/
public class NthNode {
    public static void main(String[] args) {
        List list = new List();
        int n = 3;
        for (int i = 1; i <= 5; i++) {
            list.addLast(new Node(i));
        }
        list.display();
        System.out.println(n + " th node from end = " + getNthFromLast(list.getHead(), n) + "\n");
        Node node = removeNthNodeEnd(list.getHead(), n);
        list.display(node);
    }

    //Stop slow pointer at n -1 th node and move fast pointer to the last node of the list
    //this way we can make n -1 th node's next to point to n + 1 the node and remove nth node
    private static int getNthFromLast(Node head, int n) {
        int count = 1;
        Node current = head;
        while (count < n && current != null) {
            current = current.next;
            ++count;
        }
        if (current == null) {
            return -1;
        }
        Node temp = head;
        while (current != null && current.next != null) {
            temp = temp.next;
            current = current.next;
        }
        return temp.getValue();
    }

    private static Node removeNthNodeEnd(Node head, int n) {
        var fast = head;
        for (int i = 0; i < n; ++i) {
            fast = fast.next;
        }
        if (fast == null) {
            return head.next;
        }
        var slow = head;
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }
        slow.next = slow.next.next;
        return head;
    }

}
